### ISSUE TITLE
SentenceTransformerTrainer compute_metrics

### ISSUE THREAD

**Issue by Samoed** [UKPLab/sentence-transformers#2888]
Hi! I tried to train my model and evaluate it using `compute_metrics`, but I didn't get any metrics. Is this a bug, or is it not supposed to work?

Code for test:
https://colab.research.google.com/drive/11sml4nfhkVVoZy0fTsll6BLgpHYz-BWD

**Comment by ir2718** [comment]
Hi,

I think this is the expected behaviour. but I'm not 100% sure. The `compute_metrics` function usually takes in a list of predictions which you can then use to calculate the metrics. With triplet training, the difference is in the evaluation procedure, such that the evaluators take in a SentenceTransformer model used to calculate the anchor, positive, and negative embeddings, which are then used for calculating metrics. 

Besides, there is no obvious way to define false positives and false negatives with triplet information. The only thing you can say when you've got triplet data is whether a positive is closer to the anchor than the negative, hence accuracy is only calculated.

**Comment by Samoed** [comment]
There's no information about this expected behavior or not. Maybe printing warning if the function is passed will be good.

**Comment by jacklanda** [comment]
+1

**Comment by tomaarsen** [comment]
Hello!

The `compute_metrics` argument is inherited from the `transformers` `Trainer` superclass, where it is primarily used here: https://github.com/huggingface/transformers/blob/b54109c7466f6e680156fbd30fa929e2e222d730/src/transformers/trainer.py#L4184-L4192

This happens during every evaluation, but Sentence Transformer models don't usually train with single inputs & single outputs, so the `logits` and `labels` are both None here. To be precise, they are None because the `prediction_step` calls `compute_loss`, which is implemented in the `SentenceTransformerTrainer`: https://github.com/UKPLab/sentence-transformers/blob/1802076d4eae42ff0a5629e1b04e75785d4e193b/sentence_transformers/trainer.py#L345-L351

As a result, the `compute_metrics` function is never called over the `EvalPrediction`.

---

In Sentence Transformers, if you want to compute evaluations, it is recommended to use one of the [Evaluators](https://sbert.net/docs/package_reference/sentence_transformer/evaluation.html). There's some more information about them [here](https://sbert.net/docs/sentence_transformer/training_overview.html#evaluator). 

If you want to create your own evaluator, you can subclass the [SentenceEvaluator class](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/evaluation/SentenceEvaluator.py#L10-L83). You can pass an evaluator to the STTrainer, or even a list if you have multiple.

- Tom Aarsen

**Comment by Samoed** [comment]
Thank you very much! Maybe add a warning about this? Because this is a bit unexpected behavior
